### DISCUSSION TITLE
Tool Setter module / Extended Macro functionality?

### DISCUSSION THREAD

**Discussion by Durahl** [winder/Universal-G-Code-Sender#1793] [General]
### Description

After using UGCS-P for a _long_ while I decided to have a look at CNC alternatives for a much-needed productivity boost like either a Tool Setter module or the ability to create extended Macros that would enable such a feature as constantly having to whip out a touch off plate is kinda starting to get old 😭

CNCjs, the only alternative that looked the part, had the _extended_ Macro functionality that would enable a Tool Setter but was severely lacking in the usability department - My questions about, among others, custom shortcuts to enable the use of my [Stream Deck Pendant](https://youtu.be/wk1G4gGxdxE) were pretty much instantly shot down by informing me that the Project is not being maintained anymore 😑

Coming back to UGCS-P from my short _"The grass is always greener on the other side"_ I realize that the built in Updater is doing a fairly laid-back job by keeping me on 2.0 Nightly when there's apparently long since been newer versions available on GitHub? 😫

Soo yea... In high hopes after seeing a new version I scoured UGCS-P from top to bottom but could neither find a Tool Setter module nor any advancements in the Macro department that would enable such a productivity boost? 🤔

Am I missing something?

**Comment by breiler** [comment]
The updater is not properly configured yet, so no it doesn't work. =(
Should probably be hidden to avoid future confusion.

We have not implemented a tool change feature, more details on the subject can be found here: #172

The macro functionality is pretty simple allowing you to send plain gcode (with an option to add parameters). What do you mean by an extended macro functionality? Is it something like this #1158

**Reply by Durahl** [reply]
They may be going hand in hand, but I wasn't talking about a Tool Changer ( would be nice though having one? ) but a Tool Setter allowing the transfer of the first Tools Stock Touch Off location to be transferred to the subsequent Tools and their Tool Lengths which is invaluable when the first Tool/OP has removed the top of the Stock and the subsequent Tools/OPs also rely on that now gone Touch Off location.

When I say _Extended_ Macro Functionality, [I'm talking about something like these two Macros](https://github.com/cncjs/CNCjs-Macros/tree/master/C3D_BitSetter) that can be used in CNCjs for a Toolsetter.

Unless I've been missing something, UGCS-P's Macro Functionality would not allow for such Macros?
Not that I'm confident if there were any but is there any documentation on what is possible with Macros in UGCS-P?

I couldn't really give them a try because of my unfamiliarity with CNCjs but [here's a video how it is supposedly working](https://youtu.be/uAshB6v0xNY?t=1230).

**Reply by breiler** [reply]
The documentation section for macros is unfortunately empty: https://github.com/winder/Universal-G-Code-Sender/wiki/Usage#macros

There is a help-button which displays the few options you have. It can substitute some values and allows for a simple user input. But other than that it just sends plain Gcode line separated with semicolon.
![image](https://user-images.githubusercontent.com/8962024/152104611-71d08842-c173-4e7a-a616-2151b662854d.png)

Similar requests for a more advanced macros has been made here #1426 
It would be a good idea get some inspiration on how this is done in cncjs, maybe even to have interchangeable macro scripts.

**Reply by Durahl** [reply]
Mhh... Yea... UGCS-P's Macro functionality certainly doesn't look like its gonna cut it compared to what the CNCjs example Tool Setter Macro demand - Bummer.

On a somewhat unrelated issue regarding the use of a Gamepad as a Pendant... The [Gamepad Layout Table](https://github.com/winder/Universal-G-Code-Sender/wiki/Usage#gamepad-layout) uses **Multiply/Divide Jog Feed Rate** instead of **Multiply/Divide XY/Z Step Size**. Is it considered a better practice to use a static steps size with a varying feed rate when using the analog sticks of a Gamepad as opposed to a varying step size with a static feed rate when using digital buttons? I assume if this is the case then one should also be using a small - 1mm - step size for use with the Gamepad?

I'm asking because with my Stream Deck Pendant I'm having a leaning tendency towards the latter but the XY & Z axis having their Multipliers/Dividers split is kinda annoying because I have only so many buttons available on my 5x3 Stream Deck 🤔
